exec/linux/compare/spot/man/genaut.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.4.
.TH GENAUT "1" "April 2023" "genaut (spot) 2.11.5" "User Commands"
.SH NAME
genaut \- generate ω-automata from scalable patterns
.SH SYNOPSIS
.B genaut
[\fI\,OPTION\/\fR...]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Generate ω\-automata from predefined patterns.
.SS "Pattern selection:"
.TP
\fB\-\-cyclist\-proof\-dba\fR=\fI\,RANGE\/\fR
A DBA with N+2 states that should be included
in cyclist\-trace\-nba=B.
.TP
\fB\-\-cyclist\-trace\-nba\fR=\fI\,RANGE\/\fR
An NBA with N+2 states that should include
cyclist\-proof\-dba=B.
.TP
\fB\-\-ks\-nca\fR=\fI\,RANGE\/\fR
A co\-Büchi automaton with 2N+1 states for which
any equivalent deterministic co\-Büchi automaton
has at least 2^N/(2N+1) states.
.TP
\fB\-\-l\-dsa\fR=\fI\,RANGE\/\fR
A deterministic Streett automaton with 4N states
with no equivalent deterministic Rabin automaton
of less than N! states.
.TP
\fB\-\-l\-nba\fR=\fI\,RANGE\/\fR
A Büchi automaton with 3N+1 states whose
complementary Streett automaton needs at least N!
states.
.TP
\fB\-\-m\-nba\fR=\fI\,RANGE\/\fR
An NBA with N+1 states whose determinization needs
at least N! states
.PP
RANGE may have one of the following forms: 'INT', 'INT..INT', or '..INT'.
In the latter case, the missing number is assumed to be 1.
.SS "Output format:"
.TP
\fB\-8\fR, \fB\-\-utf8\fR
enable UTF\-8 characters in output (ignored with
\fB\-\-lbtt\fR or \fB\-\-spin\fR)
.TP
\fB\-\-check\fR[=\fI\,PROP\/\fR]
test for the additional property PROP and output
the result in the HOA format (implies \fB\-H\fR).  PROP
may be some prefix of 'all' (default),
\&'unambiguous', 'stutter\-invariant',
\&'stutter\-sensitive\-example', 'semi\-determinism',
or 'strength'.
.TP
\fB\-d\fR, \fB\-\-dot\fR[=\fI\,1\/\fR|\:\fI\,a\/\fR|\:\fI\,A\/\fR|\:\fI\,b\/\fR|\:\fI\,B\/\fR|\:\fI\,c\/\fR|\:\fI\,C(COLOR)\/\fR|\:\fI\,e\/\fR|\:\fI\,E\/\fR|\:\fI\,f(FONT)\/\fR|\:\fI\,h\/\fR|\:\fI\,i(ID)\/\fR|\:\fI\,k\/\fR|\:\fI\,K\/\fR|\:\fI\,n\/\fR|\:\fI\,N\/\fR|\:\fI\,o\/\fR|\:\fI\,r\/\fR|\:\fI\,R\/\fR|\:\fI\,s\/\fR|\:\fI\,t\/\fR|\:\fI\,u\/\fR|\:\fI\,v\/\fR|\:\fI\,y\/\fR|\:\fI\,+INT\/\fR|\:\fI\,<INT\/\fR|\:\fI\,#\/\fR]
GraphViz's format.  Add letters for (1) force
numbered states, (a) show acceptance condition
(default), (A) hide acceptance condition, (b)
acceptance sets as bullets, (B) bullets except for
Büchi/co\-Büchi automata, (c) force circular
nodes, (C) color nodes with COLOR, (d) show
origins when known, (e) force elliptic nodes, (E)
force rEctangular nodes, (f(FONT)) use FONT, (g)
hide edge labels, (h) horizontal layout, (i) or
(i(GRAPHID)) add IDs, (k) use state labels when
possible, (K) use transition labels (default), (n)
show name, (N) hide name, (o) ordered transitions,
(r) rainbow colors for acceptance sets, (R) color
acceptance sets by Inf/Fin, (s) with SCCs, (t)
force transition\-based acceptance, (u) hide true
states, (v) vertical layout, (y) split universal
edges by color, (+INT) add INT to all set numbers,
(<INT) display at most INT states, (#) show
internal edge numbers
.TP
\fB\-H\fR, \fB\-\-hoaf\fR[=\fI\,1.1\/\fR|\:\fI\,i\/\fR|\:\fI\,k\/\fR|\:\fI\,l\/\fR|\:\fI\,m\/\fR|\:\fI\,s\/\fR|\:\fI\,t\/\fR|\:\fI\,v\/\fR]
Output the automaton in HOA format
(default).  Add letters to select (1.1) version
1.1 of the format, (i) use implicit labels for
complete deterministic automata, (s) prefer
state\-based acceptance when possible [default],
(t) force transition\-based acceptance, (m) mix
state and transition\-based acceptance, (k) use
state labels when possible, (l) single\-line
output, (v) verbose properties
.TP
\fB\-\-lbtt\fR[=\fI\,t\/\fR]
LBTT's format (add =t to force transition\-based
acceptance even on Büchi automata)
.TP
\fB\-\-name\fR=\fI\,FORMAT\/\fR
set the name of the output automaton
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,FORMAT\/\fR
send output to a file named FORMAT instead of
standard output.  The first automaton sent to a
file truncates it unless FORMAT starts with '>>'.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
suppress all normal output
.TP
\fB\-s\fR, \fB\-\-spin\fR[=\fI\,6\/\fR|\:\fI\,c\/\fR]
Spin neverclaim (implies \fB\-\-ba\fR).  Add letters to
select (6) Spin's 6.2.4 style, (c) comments on
states
.TP
\fB\-\-stats\fR=\fI\,FORMAT\/\fR, \fB\-\-format\fR=\fI\,FORMAT\/\fR
output statistics about the automaton
.SS "Any FORMAT string may use the following interpreted sequences:"
.TP
%%
a single %
.TP
%a
number of acceptance sets
.TP
%c, %[LETTERS]c
number of SCCs; you may filter the SCCs to count
using the following LETTERS, possibly
concatenated: (a) accepting, (r) rejecting, (c)
complete, (v) trivial, (t) terminal, (w) weak,
(iw) inherently weak. Use uppercase letters to
negate them.
.TP
%d
1 if the output is deterministic, 0 otherwise
.TP
%e, %[LETTER]e
number of edges (add one LETTER to select (r)
reachable [default], (u) unreachable, (a) all).
.TP
%F
the name of the pattern
.TP
%g, %[LETTERS]g
acceptance condition (in HOA syntax); add brackets
to print an acceptance name instead and LETTERS to
tweak the format: (0) no parameters, (a)
accentuated, (b) abbreviated, (d) style used in
dot output, (g) no generalized parameter, (l)
recognize Street\-like and Rabin\-like, (m) no main
parameter, (p) no parity parameter, (o) name
unknown acceptance as 'other', (s) shorthand for
\&'lo0'.
.TP
%h
the automaton in HOA format on a single line (use
%[opt]h to specify additional options as in
\fB\-\-hoa\fR=\fI\,opt)\/\fR
.TP
%L
the argument of the pattern
.TP
%m
name of the automaton
.TP
%n
number of nondeterministic states in output
.TP
%p
1 if the output is complete, 0 otherwise
.TP
%r
wall\-clock time elapsed in seconds (excluding
parsing)
.TP
%R, %[LETTERS]R
CPU time (excluding parsing), in seconds; Add
LETTERS to restrict to(u) user time, (s) system
time, (p) parent process, or (c) children
processes.
.TP
%s, %[LETTER]s
number of states (add one LETTER to select (r)
reachable [default], (u) unreachable, (a) all).
.TP
%t, %[LETTER]t
number of transitions (add one LETTER to select
(r) reachable [default], (u) unreachable, (a)
all).
.TP
%u, %[e]u
number of states (or [e]dges) with universal
branching
.TP
%u, %[LETTER]u
1 if the automaton contains some universal
branching (or a number of [s]tates or [e]dges with
universal branching)
.TP
%w
one word accepted by the output automaton
.TP
%x, %[LETTERS]x
number of atomic propositions declared in the
automaton;  add LETTERS to list atomic
propositions with (n) no quoting, (s) occasional
double\-quotes with C\-style escape, (d)
double\-quotes with C\-style escape, (c)
double\-quotes with CSV\-style escape, (p) between
parentheses, any extra non\-alphanumeric character
will be used to separate propositions
.SS "Miscellaneous options:"
.TP
\fB\-\-help\fR
print this help
.TP
\fB\-\-version\fR
print program version
.PP
Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.
.SH BIBLIOGRAPHY
Prefixes used in pattern names refer to the following papers:
.TP
ks
D. Kuperberg, M. Skrzypczak: On Determinisation of Good-for-Games
Automata.  Proceedings of ICALP'15.
.TP
l
C. Löding: Optimal Bounds for Transformations of ω-Automata.
Proceedings of FSTTCS'99.
.TP
m
M. Michel: Complementation is more difficult with automata on
infinite words.  CNET, Paris (1988).  Unpublished manuscript.
.SH "REPORTING BUGS"
Report bugs to <spot@lrde.epita.fr>.
.SH COPYRIGHT
Copyright \(co 2023 Laboratoire de Recherche de l'Epita (LRE)
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR autfilt (1),
.BR genltl (1),
.BR randaut (1),
.BR randltl (1)


exec/linux/compare/spot-new/genaut.cc
// -*- coding: utf-8 -*-
// Copyright (C) 2017-2019, 2022-2023 Laboratoire de Recherche et
// Développement de l'Epita (LRDE).
//
// This file is part of Spot, a model checking library.
//
// Spot is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// Spot is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
// License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include "common_sys.hh"

#include <iostream>
#include <fstream>
#include <argp.h>
#include <cstdlib>
#include "error.h"
#include <vector>

#include "common_setup.hh"
#include "common_aoutput.hh"
#include "common_range.hh"
#include "common_cout.hh"

#include <cassert>
#include <iostream>
#include <sstream>
#include <set>
#include <string>
#include <cmath>
#include <cstdlib>
#include <cstring>
#include <spot/gen/automata.hh>

using namespace spot;

static const char argp_program_doc[] =
  "Generate ω-automata from predefined patterns.";

static const argp_option options[] =
  {
    /**************************************************/
    // Keep this alphabetically sorted (expect for aliases).
    { nullptr, 0, nullptr, 0, "Pattern selection:", 1},
    { "ks-nca", gen::AUT_KS_NCA, "RANGE", 0,
      "A co-Büchi automaton with 2N+1 states for which any equivalent "
      "deterministic co-Büchi automaton has at least 2^N/(2N+1) states.", 0},
    { "l-nba", gen::AUT_L_NBA, "RANGE", 0,
      "A Büchi automaton with 3N+1 states whose complementary Streett "
      "automaton needs at least N! states.", 0},
    { "l-dsa", gen::AUT_L_DSA, "RANGE", 0,
      "A deterministic Streett automaton with 4N states with no "
      "equivalent deterministic Rabin automaton of less than N! states.", 0},
    { "m-nba", gen::AUT_M_NBA, "RANGE", 0,
      "An NBA with N+1 states whose determinization needs at least "
      "N! states", 0},
    { "cyclist-trace-nba", gen::AUT_CYCLIST_TRACE_NBA, "RANGE", 0,
      "An NBA with N+2 states that should include cyclist-proof-dba=B.", 0},
    { "cyclist-proof-dba", gen::AUT_CYCLIST_PROOF_DBA, "RANGE", 0,
      "A DBA with N+2 states that should be included "
      "in cyclist-trace-nba=B.", 0},
    RANGE_DOC,
  /**************************************************/
    { nullptr, 0, nullptr, 0, "Miscellaneous options:", -1 },
    { nullptr, 0, nullptr, 0, nullptr, 0 }
  };

struct job
{
  gen::aut_pattern_id pattern;
  struct range range;
};

typedef std::vector<job> jobs_t;
static jobs_t jobs;

const struct argp_child children[] =
  {
    { &aoutput_argp, 0, nullptr, 3 },
    { &aoutput_o_format_argp, 0, nullptr, 4 },
    { &misc_argp, 0, nullptr, -1 },
    { nullptr, 0, nullptr, 0 }
  };

static void
enqueue_job(int pattern, const char* range_str)
{
  job j;
  j.pattern = static_cast<gen::aut_pattern_id>(pattern);
  j.range = parse_range(range_str);
  jobs.push_back(j);
}

static int
parse_opt(int key, char* arg, struct argp_state*)
{
  // Called from C code, so should not raise any exception.
  BEGIN_EXCEPTION_PROTECT;
  if (key >= gen::AUT_BEGIN && key < gen::AUT_END)
    {
      enqueue_job(key, arg);
      return 0;
    }
  END_EXCEPTION_PROTECT;
  return ARGP_ERR_UNKNOWN;
}

static void
output_pattern(gen::aut_pattern_id pattern, int n)
{
  process_timer timer;
  timer.start();
  twa_graph_ptr aut = spot::gen::aut_pattern(pattern, n);
  timer.stop();
  automaton_printer printer;
  printer.print(aut, timer, nullptr, aut_pattern_name(pattern), n);
}

static void
run_jobs()
{
  for (const auto& j: jobs)
    {
      int inc = (j.range.max < j.range.min) ? -1 : 1;
      int n = j.range.min;
      for (;;)
        {
          output_pattern(j.pattern, n);
          if (n == j.range.max)
            break;
          n += inc;
        }
    }
}


int
main(int argc, char** argv)
{
  return protected_main(argv, [&] {
      strcpy(F_doc, "the name of the pattern");
      strcpy(L_doc, "the argument of the pattern");

      const argp ap = { options, parse_opt, nullptr, argp_program_doc,
                        children, nullptr, nullptr };

      if (int err = argp_parse(&ap, argc, argv, ARGP_NO_HELP, nullptr, nullptr))
        exit(err);

      if (jobs.empty())
        error(1, 0, "Nothing to do.  Try '%s --help' for more information.",
              program_name);

      run_jobs();
      flush_cout();
      return 0;
    });
}
